### PR TITLE
feat(dashboards): Adds pagination and table sort support to issue widget viewer

### PR DESCRIPTION
**Pull request by edwardgou-sentry**
Extends the Widget Viewer pagination and column sort functionality to also support Issue Widgets.

### CODE REVIEW

File: static/app/components/modals/widgetViewerModal/widgetViewerTableCell.tsx [hunk 1]
```diff
@@ -37,7 +39,32 @@ type Props = {
...
+      <SortLink
+        align={align}
+        title={<StyledTooltip title={column.name}>{column.name}</StyledTooltip>}
+        direction={widget.queries[0].orderby === sortField ? 'desc' : undefined}
+        canSort
+        generateSortLink={() => {
+          return sortField
+            ? {
+                ...location,
+                query: {...location.query, [WidgetViewerQueryField.SORT]: sortField},
```
Comment (narsaynorath) [inline static/app/components/modals/widgetViewerModal/widgetViewerTableCell.tsx:59]: Should we be negating the sort here if we are sorting by this field? Currently I'm unable to reverse the sorting.

File: static/app/components/modals/widgetViewerModal/widgetViewerTableCell.tsx [hunk 2]
```diff
@@ -37,7 +39,32 @@ type Props = {
...
+    const tableMeta = tableData?.meta;
+    const align = fieldAlignment(column.name, column.type, tableMeta);
+    const sortField = getSortField(String(column.key));
+
+    return (
+      <SortLink
+        align={align}
+        title={<StyledTooltip title={column.name}>{column.name}</StyledTooltip>}
+        direction={widget.queries[0].orderby === sortField ? 'desc' : undefined}
+        canSort
```
Comment (narsaynorath) [inline static/app/components/modals/widgetViewerModal/widgetViewerTableCell.tsx:54]: Can we use `sortField` here to set `canSort={!!sortField}` so we don't have to do the ternary in `generateSortLink`?

From what I understand if you can't sort, then it won't render with a clickable header

Comment (edwardgou-sentry) [reply static/app/components/modals/widgetViewerModal/widgetViewerTableCell.tsx]: Yup, updated canSort and generateSortLink!

File: static/app/components/modals/widgetViewerModal/widgetViewerTableCell.tsx (hunk 1 above)
Comment (edwardgou-sentry) [reply static/app/components/modals/widgetViewerModal/widgetViewerTableCell.tsx]: Hey Nar, Issues API doesn't support sorting in ascending order at this moment 😞 .
The api only supports a limited set of options, namely descending Events, descending Users, First Seen, and Last Seen (none of these in reverse order).
Due to this, we are only able to allow limited sorting for Issue Widgets in the Viewer.

File: static/app/components/modals/widgetViewerModal.tsx [hunk 3]
```diff
@@ -150,27 +154,49 @@ function WidgetViewerModal(props: Props) {
...
+                          ...props,
+                          widget: tableWidget,
+                        }),
+                      }}
+                      location={location}
+                    />
+                    <StyledPagination
+                      pageLinks={pageLinks}
+                      onCursor={(nextCursor, _path, _query, delta) => {
+                        let nextPage: number | undefined = isNaN(pagination.page)
```
Comment (priscilawebdev) [inline static/app/components/modals/widgetViewerModal.tsx:185]: can `nextPage` get also an `undefined` value?

File: tests/js/spec/components/modals/widgetViewerModal.spec.tsx [hunk 4]
```diff
@@ -432,5 +471,60 @@ describe('Modals -> WidgetViewerModal', function () {
...
+
+    it('sorts table when a sortable column header is clicked', function () {
+      userEvent.click(screen.getByText('events'));
+      expect(initialData.router.push).toHaveBeenCalledWith({
+        query: {modalSort: 'freq'},
+      });
+      // Need to manually set the new router location and rerender to simulate the sortable column click
+      initialData.router.location.query = {modalSort: ['freq']};
+      rerender(
+        <div style={{padding: space(4)}}>
```
Comment (priscilawebdev) [inline tests/js/spec/components/modals/widgetViewerModal.spec.tsx:483]: do we need this div element?

**Review by narsaynorath — APPROVED** [review]
Looks good to me

One question I had about the sorting though, does it make sense to highlight the LifetimeEvents/LifetimeUsers headers if it's only doing freq and user? e.g. here's a screenshot where it's showing it's descending on LifetimeEvents but it's not actually doing that
Screen Shot 2022-03-08 at 9 42 35 AM

File: tests/js/spec/components/modals/widgetViewerModal.spec.tsx (hunk 4 above)
Comment (edwardgou-sentry) [reply tests/js/spec/components/modals/widgetViewerModal.spec.tsx]: Not needed! Updated!

File: static/app/components/modals/widgetViewerModal.tsx (hunk 3 above)
Comment (edwardgou-sentry) [reply static/app/components/modals/widgetViewerModal.tsx]: Shouldn't be possible to get an `undefined` value here, I've updated the type to reflect this!

**Comment by edwardgou-sentry** [comment]
> Looks good to me
> 
> One question I had about the sorting though, does it make sense to highlight the LifetimeEvents/LifetimeUsers headers if it's only doing freq and user? e.g. here's a screenshot where it's showing it's descending on LifetimeEvents but it's not actually doing that

Good question, I added the freq and user sort fields to LifetimeEvents and LifetimeUsers since they were somewhat related, but I think it might cause some confusion for users since it's not directly sorting the Lifetime columns, as you pointed out. I've updated to remove it for now and will consult with Jasmin to see what we want to do.

**Review by priscilawebdev — APPROVED** [review]
lgtm

**Pull request merged**